File: vc-verifier/java/src/main/java/io/mosip/vercred/signature/impl/ED25519SignatureVerifierImpl.java
package io.mosip.vercred.signature.impl;

import io.mosip.vercred.contant.CredentialVerifierConstants;
import io.mosip.vercred.exception.SignatureVerificationException;
import io.mosip.vercred.signature.SignatureVerifier;

import java.security.*;

public class ED25519SignatureVerifierImpl implements SignatureVerifier {
    @Override
    public Boolean verify(PublicKey publicKey, byte[] signData, byte[] signature) {
        try {
            Signature ed25519Signature = Signature.getInstance(CredentialVerifierConstants.ED25519_ALGORITHM);
            ed25519Signature.initVerify(publicKey);
            ed25519Signature.update(signData);
            return ed25519Signature.verify(signature);
        } catch (NoSuchAlgorithmException | InvalidKeyException | SignatureException e) {
            throw new SignatureVerificationException("Error while doing signature verification using ED25519 algorithm");
        }
    }
}

File: vc-verifier/kotlin/vcverifier/src/main/java/io/mosip/vercred/vcverifier/signature/impl/ED25519SignatureVerifierImpl.kt
package io.mosip.vercred.vcverifier.signature.impl

import android.annotation.TargetApi
import android.os.Build
import io.mosip.vercred.vcverifier.signature.SignatureVerifier
import io.mosip.vercred.vcverifier.constants.CredentialVerifierConstants
import io.mosip.vercred.vcverifier.exception.SignatureVerificationException
import net.i2p.crypto.eddsa.EdDSAEngine
import org.bouncycastle.jce.provider.BouncyCastleProvider
import java.security.PublicKey
import java.security.Signature

class ED25519SignatureVerifierImpl : SignatureVerifier {


    @TargetApi(Build.VERSION_CODES.TIRAMISU)
    override fun verify(publicKey: PublicKey, signData: ByteArray, signature: ByteArray, provider: BouncyCastleProvider): Boolean {
        try {
//            val ed25519Signature =
//                Signature.getInstance(CredentialVerifierConstants.ED25519_ALGORITHM, provider)
//                    /*.apply {
//                    initVerify(publicKey)
//                    update(signData)
//                    verify(signature)
//                }*/
            val eddsaEngine = EdDSAEngine()
            eddsaEngine.initVerify(publicKey)

            // Step 3: Update the engine with the data to verify
            eddsaEngine.update(signData)

            // Step 4: Verify the signature
            return eddsaEngine.verify(signature)

        } catch (e: Exception) {
            throw SignatureVerificationException("Error while doing signature verification using ED25519 algorithm")
        }
    }
}